modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/LazyMuleArtifactContext.java

```diff
@@ -211,8 +211,49 @@ protected void registerErrors(ArtifactAst artifactAst) {
   @Override
   protected void prepareBeanFactory(ConfigurableListableBeanFactory beanFactory) {
     super.prepareBeanFactory(beanFactory);
-    trackingPostProcessor = new TrackingPostProcessor();
-    addBeanPostProcessors(beanFactory, trackingPostProcessor);
+    currentComponentInitializationState.registerTrackingPostProcessor(beanFactory);
+  }
+
+  @Override
+  public Object getBean(String name) throws BeansException {
+    try {
+      return super.getBean(name);
+    } catch (NoSuchBeanDefinitionException e) {
+      // It is possible that getting the bean failed because the bean definition was not yet registered.
+      // We will try to do that now.
+      // If it fails, we will re-throw the original exception as if the lazy initialization attempt never occurred.
+      return tryInitializeAndGetBean(name).orElseThrow(() -> e);
+    }
+  }
+
+  private Optional<Object> tryInitializeAndGetBean(String name) {
+    if (initializingBeanLock.isHeldByCurrentThread()) {
+      // Prevents reentrancy. We shouldn't try to initialize some missing bean while we are trying to initialize some other.
+      // (Dependencies should be handled by the lazy initialization mechanism without going through here)
```

Comment:
Didn't think of that, it could make for a good test. So far (AFAIK) configs cannot depend dynamically on other configs, but I could build a case with a dynamic flow-ref to a flow with an operation which dynamically references a config.

Reply:
I replaced the `Lock` with a regular synchronized block to allow reentrancy and added a `ThreadLocal` `Set` to keep track of beans that are in the process of being dynamically initialized.